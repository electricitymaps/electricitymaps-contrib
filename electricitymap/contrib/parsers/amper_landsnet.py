from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.exceptions import ParserException

SOURCE = "amper.landsnet.is"
SOURCE_URL = "https://amper.landsnet.is/generation/api/Values"


def fetch_production(
    zone_key: ZoneKey = ZoneKey("IS"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known production mix (in MW) of a given country."""
    r = session or Session()
    if target_datetime is not None:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    res = r.get(SOURCE_URL)
    if not res.ok:
        raise ParserException(
            parser="amper_landsnet.py",
            message=f"Failed to fetch {SOURCE}: {res.status_code}, err: {res.text}",
            zone_key=zone_key,
        )

    obj = res.json()
    mix = ProductionMix(
        geothermal=obj["geothermal"],
        hydro=obj["hydro"],
        oil=obj["oil"],
    )
    breakdowns = ProductionBreakdownList(logger=logger)
    breakdowns.append(
        zoneKey=zone_key,
        datetime=datetime.fromisoformat(obj["timestamp"]).replace(
            tzinfo=ZoneInfo("Atlantic/Reykjavik")
        ),
        production=mix,
        source=SOURCE,
    )
    return breakdowns.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
