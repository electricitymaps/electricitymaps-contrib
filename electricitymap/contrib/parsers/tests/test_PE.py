import json
from datetime import datetime
from pathlib import Path

from requests_mock import POST

from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.PE import API_ENDPOINT, fetch_production

# Base path for mock data files
MOCK_DATA_DIR = Path(__file__).parent / "mocks" / "PE"


def test_fetch_production_with_target_datetime(adapter, session, snapshot):
    """Test production data parsing for specific target datetime 2025-09-10."""

    def mock_response(request, context):
        """Mock response based on the request data."""
        # Parse the request data to determine which date is being requested
        form_data = request.text

        # Extract fechaInicial from the form data
        if (
            "fechaInicial=09%2F09%2F2025" in form_data
            or "fechaInicial=09/09/2025" in form_data
        ):
            # Request for 2025-09-09 (yesterday)
            mock_file = MOCK_DATA_DIR / "response_20250909.json"
        elif (
            "fechaInicial=10%2F09%2F2025" in form_data
            or "fechaInicial=10/09/2025" in form_data
        ):
            # Request for 2025-09-10 (target date)
            mock_file = MOCK_DATA_DIR / "response_20250910.json"
        else:
            # Default fallback
            mock_file = MOCK_DATA_DIR / "response_20250910.json"

        # Load and return the appropriate mock data
        with open(mock_file, encoding="utf-8") as f:
            return json.load(f)

    # Register the mock adapter
    adapter.register_uri(
        POST,
        API_ENDPOINT,
        json=mock_response,
    )

    # Test with target_datetime = 2025-09-10
    target_datetime = datetime(2025, 9, 10, 12, 0, 0)
    result = fetch_production(
        zone_key=ZoneKey("PE"),
        session=session,
        target_datetime=target_datetime,
    )

    # Verify the result using snapshot testing
    assert snapshot == result


def test_fetch_production_data_structure(adapter, session):
    """Test that the production data has the expected structure and values."""

    def mock_response(request, context):
        """Mock response that returns consistent data."""
        mock_file = MOCK_DATA_DIR / "response_20250910.json"
        with open(mock_file, encoding="utf-8") as f:
            return json.load(f)

    adapter.register_uri(
        POST,
        API_ENDPOINT,
        json=mock_response,
    )

    target_datetime = datetime(2025, 9, 10, 12, 0, 0)
    result = fetch_production(
        zone_key=ZoneKey("PE"),
        session=session,
        target_datetime=target_datetime,
    )

    # Basic structure checks
    assert isinstance(result, list)
    assert len(result) > 0

    # Check first production breakdown
    first_item = result[0]
    assert "zoneKey" in first_item
    assert "datetime" in first_item
    assert "production" in first_item
    assert "source" in first_item

    # Check that we have the expected energy sources
    production = first_item["production"]
    expected_sources = {"biomass", "hydro", "gas", "solar", "wind", "coal", "oil"}

    # Check that at least some expected sources are present
    production_sources = set(production.keys())
    assert len(production_sources.intersection(expected_sources)) > 0

    # Check that all values are numeric
    for _, value in production.items():
        assert isinstance(value, int | float)
        assert value >= 0  # Production values should be non-negative
