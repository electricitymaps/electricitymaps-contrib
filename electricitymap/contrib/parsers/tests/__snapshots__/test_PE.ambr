# serializer version: 1
# name: test_fetch_production
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 39.7,
        'coal': 0.0,
        'gas': 1581.632,
        'hydro': 4687.987,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 141.694,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 0, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 39.0,
        'coal': 0.0,
        'gas': 1550.105,
        'hydro': 4505.311,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 133.901,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 1, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 41.1,
        'coal': 0.0,
        'gas': 1563.946,
        'hydro': 4542.161,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 116.896,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 1, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 41.6,
        'coal': 0.0,
        'gas': 1580.377,
        'hydro': 4418.322,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 107.936,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 2, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 42.1,
        'coal': 0.0,
        'gas': 1577.064,
        'hydro': 4432.829,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 112.561,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 2, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 43.0,
        'coal': 0.0,
        'gas': 1566.439,
        'hydro': 4382.714,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 120.539,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 3, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 41.4,
        'coal': 0.0,
        'gas': 1565.062,
        'hydro': 4287.968,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 96.796,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 3, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 43.0,
        'coal': 0.0,
        'gas': 1575.853,
        'hydro': 4395.27,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 79.518,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 4, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 42.0,
        'coal': 0.0,
        'gas': 1561.994,
        'hydro': 4275.16,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 58.002,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 4, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 40.9,
        'coal': 0.0,
        'gas': 1613.413,
        'hydro': 4397.272,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 47.394,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 5, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 41.7,
        'coal': 0.0,
        'gas': 1773.063,
        'hydro': 4395.052,
        'oil': 0.0,
        'solar': 0.423,
        'wind': 32.464,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 5, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 42.8,
        'coal': 0.0,
        'gas': 1907.137,
        'hydro': 4418.248,
        'oil': 0.0,
        'solar': 11.556,
        'wind': 36.177,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 6, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 43.4,
        'coal': 0.0,
        'gas': 1903.247,
        'hydro': 4375.145,
        'oil': 0.0,
        'solar': 84.361,
        'wind': 44.775,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 6, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 42.6,
        'coal': 0.0,
        'gas': 1905.211,
        'hydro': 4461.667,
        'oil': 0.0,
        'solar': 190.985,
        'wind': 32.221,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 7, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 42.5,
        'coal': 0.0,
        'gas': 1942.96,
        'hydro': 4517.466,
        'oil': 0.0,
        'solar': 279.024,
        'wind': 27.76,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 7, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 42.7,
        'coal': 0.0,
        'gas': 2328.21,
        'hydro': 4615.279,
        'oil': 0.0,
        'solar': 320.55,
        'wind': 25.17,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 8, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 42.5,
        'coal': 0.0,
        'gas': 2454.929,
        'hydro': 4503.523,
        'oil': 0.0,
        'solar': 347.672,
        'wind': 26.262,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 8, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 36.8,
        'coal': 0.0,
        'gas': 2625.72,
        'hydro': 4531.692,
        'oil': 0.0,
        'solar': 359.147,
        'wind': 30.431,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 9, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 32.0,
        'coal': 0.0,
        'gas': 2800.437,
        'hydro': 4620.383,
        'oil': 0.0,
        'solar': 376.286,
        'wind': 58.38,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 9, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 32.6,
        'coal': 0.0,
        'gas': 2822.613,
        'hydro': 4553.179,
        'oil': 0.0,
        'solar': 382.128,
        'wind': 81.439,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 10, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 30.7,
        'coal': 0.0,
        'gas': 2922.659,
        'hydro': 4592.594,
        'oil': 0.0,
        'solar': 383.337,
        'wind': 94.942,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 10, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 31.6,
        'coal': 0.0,
        'gas': 3018.674,
        'hydro': 4642.007,
        'oil': 0.0,
        'solar': 339.944,
        'wind': 86.13,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 11, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 30.4,
        'coal': 0.0,
        'gas': 2961.89,
        'hydro': 4628.321,
        'oil': 0.0,
        'solar': 357.751,
        'wind': 121.068,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 11, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 30.8,
        'coal': 0.0,
        'gas': 3052.03,
        'hydro': 4512.078,
        'oil': 0.0,
        'solar': 324.771,
        'wind': 144.539,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 12, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 30.0,
        'coal': 0.0,
        'gas': 2992.392,
        'hydro': 4477.638,
        'oil': 0.0,
        'solar': 346.854,
        'wind': 163.701,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 12, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 31.1,
        'coal': 0.0,
        'gas': 2945.642,
        'hydro': 4510.965,
        'oil': 0.0,
        'solar': 208.21,
        'wind': 259.323,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 13, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 29.1,
        'coal': 0.0,
        'gas': 3005.798,
        'hydro': 4548.965,
        'oil': 0.0,
        'solar': 196.803,
        'wind': 210.576,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 13, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 29.6,
        'coal': 0.0,
        'gas': 2932.877,
        'hydro': 4501.439,
        'oil': 0.0,
        'solar': 338.772,
        'wind': 210.126,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 14, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 29.5,
        'coal': 0.0,
        'gas': 2998.214,
        'hydro': 4522.273,
        'oil': 0.0,
        'solar': 348.052,
        'wind': 207.732,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 14, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 26.4,
        'coal': 0.0,
        'gas': 2930.015,
        'hydro': 4490.295,
        'oil': 41.0,
        'solar': 352.953,
        'wind': 241.071,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 15, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 28.3,
        'coal': 0.0,
        'gas': 3037.318,
        'hydro': 4339.493,
        'oil': 88.4,
        'solar': 294.734,
        'wind': 256.171,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 15, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 28.1,
        'coal': 0.0,
        'gas': 3068.125,
        'hydro': 4294.94,
        'oil': 91.8,
        'solar': 241.186,
        'wind': 241.739,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 16, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 28.2,
        'coal': 0.0,
        'gas': 3042.141,
        'hydro': 4406.698,
        'oil': 94.5,
        'solar': 178.134,
        'wind': 245.038,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 16, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 29.9,
        'coal': 0.0,
        'gas': 3034.953,
        'hydro': 4338.674,
        'oil': 15.4,
        'solar': 115.775,
        'wind': 232.69,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 17, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 30.7,
        'coal': 0.0,
        'gas': 2962.28,
        'hydro': 4438.303,
        'oil': 0.0,
        'solar': 59.372,
        'wind': 188.954,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 17, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 32.0,
        'coal': 0.0,
        'gas': 2943.25,
        'hydro': 4352.437,
        'oil': 0.0,
        'solar': 17.418,
        'wind': 164.885,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 18, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 32.0,
        'coal': 0.0,
        'gas': 2920.596,
        'hydro': 4392.21,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 179.126,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 18, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 32.4,
        'coal': 0.0,
        'gas': 3067.697,
        'hydro': 4560.945,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 165.55,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 19, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 33.5,
        'coal': 0.0,
        'gas': 2996.269,
        'hydro': 4731.293,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 183.6,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 19, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 33.8,
        'coal': 0.0,
        'gas': 3009.521,
        'hydro': 4695.945,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 210.7,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 20, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 33.6,
        'coal': 0.0,
        'gas': 2996.916,
        'hydro': 4679.196,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 232.4,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 20, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 33.2,
        'coal': 0.0,
        'gas': 3110.768,
        'hydro': 4482.913,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 261.8,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 21, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 33.4,
        'coal': 0.0,
        'gas': 3028.597,
        'hydro': 4544.399,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 219.8,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 21, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 33.2,
        'coal': 0.0,
        'gas': 2846.725,
        'hydro': 4567.491,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 257.0,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 22, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 37.3,
        'coal': 0.0,
        'gas': 2775.06,
        'hydro': 4597.061,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 241.0,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 22, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 41.7,
        'coal': 0.0,
        'gas': 2598.62,
        'hydro': 4598.875,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 241.85,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 23, 0, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 42.5,
        'coal': 0.0,
        'gas': 2528.327,
        'hydro': 4628.935,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 207.692,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 5, 23, 30, tzinfo=zoneinfo.ZoneInfo(key='America/Lima')),
      'production': dict({
        'biomass': 42.0,
        'coal': 0.0,
        'gas': 1714.83,
        'hydro': 4516.703,
        'oil': 0.0,
        'solar': 0.0,
        'wind': 204.447,
      }),
      'source': 'coes.org.pe',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PE',
    }),
  ])
# ---
